mongocrypt/build.gradle.kts

```diff
@@ -0,0 +1,188 @@
+
+val jnaMapping: Map<String, String> = mapOf(
+    "rhel-62-64-bit" to "linux-x86-64",
+    "rhel72-zseries-test" to "linux-s390x",
+    "rhel-71-ppc64el" to "linux-ppc64le",
+    "ubuntu1604-arm64" to "linux-aarch64",
+    "windows-test" to "win32-x86-64",
+    "macos" to "darwin"
+)
+
+sourceSets {
+    main {
+        java {
+            resources {
+                srcDirs(jnaResourcesDir)
+            }
+        }
+    }
+}
+
+tasks.register<Download>("downloadJava") {
+    src(downloadUrl)
+    dest("${jnaDownloadsDir}/$localBinariesArchiveName")
+    overwrite(true)
+    /* To make sure we don't download archive with binaries if it hasn't been changed in S3 bucket since last download.*/
+    onlyIfModified(true)
+}
+
+tasks.register<Copy>("unzipJava") {
+    /*
+        Clean up the directory first if the task is not UP-TO-DATE.
+        This can happen if the download revision has been changed and the archive is downloaded again.
+     */
+    doFirst {
+        println("Cleaning up $jnaResourcesDir")
+        delete(jnaResourcesDir)
+    }
+    from(tarTree(resources.gzip("${jnaDownloadsDir}/$localBinariesArchiveName")))
+    include(jnaMapping.keys.flatMap {
+        listOf("${it}/nocrypto/**/libmongocrypt.so", "${it}/lib/**/libmongocrypt.dylib", "${it}/bin/**/mongocrypt.dll" )
+    })
+    eachFile {
+        path = "${jnaMapping[path.substringBefore("/")]}/${name}"
+    }
+    into(jnaResourcesDir)
+    dependsOn("downloadJava")
+
+    doLast {
+        println("jna.library.path contents: \n  ${fileTree(jnaResourcesDir).files.joinToString(",\n  ")}")
+    }
+}
+
+// The `processResources` task (defined by the `java-library` plug-in) consumes files in the main source set.
+// Add a dependency on `unzipJava`. `unzipJava` adds libmongocrypt libraries to the main source set.
+tasks.processResources {
+    mustRunAfter(tasks.named("unzipJava"))
+}
+
+tasks.register("downloadJnaLibs") {
+    dependsOn("downloadJava", "unzipJava")
+}
+
+tasks.test {
+    systemProperty("jna.debug_load", "true")
+    systemProperty("jna.library.path", jnaResources)
+    useJUnitPlatform()
+    testLogging {
+        events("passed", "skipped", "failed")
+    }
+
+    doFirst {
+        println("jna.library.path contents:")
+        println(fileTree(jnaResources)  {
+            this.setIncludes(listOf("*.*"))
+        }.files.joinToString(",\n  ", "  "))
+    }
+    dependsOn("downloadJnaLibs", "downloadJava", "unzipJava")
+}
+
+tasks.withType<AbstractPublishToMaven> {
+    description = """$description
+        | System properties:
+        | =================
+        |
+        | jnaLibsPath    : Custom local JNA library path for inclusion into the build (rather than downloading from s3)
+        | gitRevision    : Optional Git Revision to download the built resources for from s3.
+    """.trimMargin()
+}
+
+tasks.jar {
+    //NOTE this enables depending on the mongocrypt from driver-core
+   dependsOn("downloadJnaLibs")
+}
+
+tasks.javadoc {
```

Comment:
Yes, this will be covered under JAVA-5584, which includes removing Javadoc generation for this module. However, there's a `MongoCryptException` that might not be moved to the internal package, so we could consider retaining the Javadoc because of this.